### PR TITLE
posix: Various additions to tmpfs

### LINKED ISSUE #217: posix: tmp_fs needs fixes in various error paths
- In `link()`, `unlink()`, `rename()`, `mkdir()`, `symlink()`, `mkdev()` and `mkfifo()` we assert that entries do not (or do) exist. We should return an error instead (`mksocket()` is correct).
- In `readSome()`, we have another assertion that should return 0 instead (and no error).

### PR DESCRIPTION
**Pull request by Dennisbonke**
This PR aims to implement `seek` and `rmdir` for the tmpfs `DirectoryNode` and `DirectoryFile`.
Python uses `rmdir()` and `rewinddir()` on directories in `/tmp` when running `python3 get-pip.py`, this PR implements the required functionality.
Furthermore, this PR changes various assertions in tmpfs to proper error returns. As such, this PR closes #217

### CODE REVIEW

File: posix/subsystem/src/tmp_fs.cpp [hunk 1]
```diff
@@ -264,7 +264,7 @@ struct DirectoryNode final : Node, std::enable_shared_from_this<DirectoryNode> {
 
 private:
 	VfsType getType() override {
-		return VfsType::directory;
+  		return VfsType::directory;
```
Comment (avdgrinten) [inline posix/subsystem/src/tmp_fs.cpp:267]: Strange indentation change.

File: posix/subsystem/src/tmp_fs.cpp [hunk 2]
```diff
@@ -292,7 +292,8 @@ struct DirectoryNode final : Node, std::enable_shared_from_this<DirectoryNode> {
 
 	async::result<frg::expected<Error, std::shared_ptr<FsLink>>> link(std::string name,
 			std::shared_ptr<FsNode> target) override {
-		assert(_entries.find(name) == _entries.end());
+		if(!(_entries.find(name) == _entries.end()))
```
Comment (avdgrinten) [inline posix/subsystem/src/tmp_fs.cpp:295]: We can use != here (same below).

File: posix/subsystem/src/tmp_fs.cpp [hunk 3]
```diff
@@ -487,7 +489,9 @@ struct Superblock final : FsSuperblock {
 
 		auto src_dir = static_cast<DirectoryNode *>(src_link->getOwner().get());
 		auto it = src_dir->_entries.find(src_link->getName());
-		assert(it != src_dir->_entries.end() && it->get() == src_link);
+		// TODO: This should really return an error, frg::expected style.
+		if(it == src_dir->_entries.end() || it->get() != src_link)
+			co_return nullptr;
```
Comment (avdgrinten) [inline posix/subsystem/src/tmp_fs.cpp:494]: We can do the conversion to `frg::expected` in this PR?

File: posix/subsystem/src/tmp_fs.cpp (hunk 2 above)
Comment (Dennisbonke) [reply posix/subsystem/src/tmp_fs.cpp:295]: This is the same as `mksocket()` uses, according to #217 that was correct

File: posix/subsystem/src/tmp_fs.cpp (hunk 3 above)
Comment (Dennisbonke) [reply posix/subsystem/src/tmp_fs.cpp]: Yes we can, done in 0699a50

File: posix/subsystem/src/main.cpp [hunk 4]
```diff
@@ -1536,8 +1536,9 @@ async::result<void> serveRequests(std::shared_ptr<Process> self,
 			auto superblock = resolver.currentLink()->getTarget()->superblock();
 			auto directory = new_resolver.currentLink()->getTarget();
 			assert(superblock == directory->superblock());
-			co_await superblock->rename(resolver.currentLink().get(),
+			auto result = co_await superblock->rename(resolver.currentLink().get(),
 					directory.get(), new_resolver.nextComponent());
+			assert(result);
```
Comment (avdgrinten) [inline posix/subsystem/src/main.cpp:1541]: Send an error to the client here.

File: posix/subsystem/src/tmp_fs.cpp [hunk 5]
```diff
@@ -307,7 +310,8 @@ struct DirectoryNode final : Node, std::enable_shared_from_this<DirectoryNode> {
 
 	async::result<frg::expected<Error>> unlink(std::string name) override {
 		auto it = _entries.find(name);
-		assert(it != _entries.end());
+		if(it == _entries.end())
+			co_return Error::alreadyExists;
```
Comment (avdgrinten) [inline posix/subsystem/src/tmp_fs.cpp:314]: The error should be `noSuchFile` here, right? (In contrast to mkdir() and friends, we are checking whether an entry *exists* and not whether it does not exist.)

**Review by avdgrinten — COMMENTED** [review]
Since the seek part is dropped from the PR, please also drop the corresponding commit via rebase.

File: posix/subsystem/src/tmp_fs.cpp [hunk 6]
```diff
@@ -316,6 +320,15 @@ struct DirectoryNode final : Node, std::enable_shared_from_this<DirectoryNode> {
 
 	async::result<frg::expected<Error, std::shared_ptr<FsLink>>> mksocket(std::string name) override;
 
+	async::result<frg::expected<Error>> rmdir(std::string name) override {
+		auto result = co_await unlink(name);
+		if(!result) {
+			std::cout << "tmpfs: Unexpected failure from unlink()" << std::endl;
+			co_return {};
```
Comment (avdgrinten) [inline posix/subsystem/src/tmp_fs.cpp:327]: Shouldn't we return an error here?

**Pull request merged**